platforms/android/tangram/src/main/java/com/mapzen/tangram/MapView.java

```diff
@@ -11,72 +12,190 @@
+     * @param handler Set the client implemented {@link HttpHandler} for retrieving remote map resources
+     *                when null {@link DefaultHttpHandler} is used
+     */
+    protected void executeMapAsyncTask(@Nullable final MapReadyCallback callback, @Nullable final HttpHandler handler) {
+        getMapAsyncTask = new MapAsyncTask(this, callback, handler);
+        getMapAsyncTask.execute();
     }
 
     /**
-     * Construct a {@code MapController}; may only be called from the UI thread
+     * Responsible to dispose any prior running getMapReadyTask
+     * Note: Can be overriden to support other ways of implementing async tasks (e.g. when using observable with RxJava)
+     */
+    protected void disposeMapReadyTask() {
+        if (getMapAsyncTask != null) {
+            getMapAsyncTask.cancel(true);
+        }
+        getMapAsyncTask = null;
+    }
+
+    /**
+     * Construct a {@code MapController} in an async thread; may only be called from the UI thread
      * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
-     * @param listener The listener to receive to receive scene load events;
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
      * the callback will be made on the UI thread
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener) {
-        return getMap(listener, null);
+    public void getMapAsync(@Nullable final MapReadyCallback readyCallback) {
+        getMapAsync(readyCallback, null);
     }
 
     /**
-     * Construct a {@code MapController}; may only be called from the UI thread
-     * @param listener The listener to receive to receive scene load events;
+     * Construct a {@code MapController} in an async thread; may only be called from the UI thread
+     * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
      * @param handler Set the client implemented {@link HttpHandler} for retrieving remote map resources
      *                when null {@link DefaultHttpHandler} is used
-     * the callback will be made on the UI thread
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener, @Nullable final HttpHandler handler) {
-        if (mapController != null) {
-            return mapController;
-        }
-        mapController = getMapInstance();
-        mapController.setSceneLoadListener(listener);
-        mapController.init(handler);
+    public void getMapAsync(@Nullable final MapReadyCallback readyCallback,
```

Comment:
Should the `MapReadyCallback` really be `Nullable`? I can't think of a case where you wouldn't want to get a reference to the `MapController`.

Reply:
Yeah makes sense will update.